design/0001-frr.md

```diff
@@ -0,0 +1,411 @@
+
+#### others
+
+There may be some common functionality in the bgp package that could be reused
+between BGP implementations. For example, the new bgp package will require
+integration with the “metrics” struct which updates prometheus in response to
+BGP events. It may be necessary to refactor these types of functionality into a
+separate package for reuse amongst BGP implementations.
+
+### Test Plan
+
+**e2e tests**: The intention is to expand on the MetalLB ‘dev-env’ KIND
+environment. After deploying this environment, it will be possible to run any
+relevant Kubernetes e2e tests (or custom MetalLB tests) against this test
+cluster. Further investigation will be required as part of [Story 1](#story-1)
+in order to determine the relevance of any Kubernetes e2e tests and if any other
+e2e tests will need to be developed.
+**unit tests**: Unit tests will be added to
+any additional code in the MetalLB repository.
+**upgrade**: We may need to add
+tests to deal with an upgrade to a newer FRR version.
+
+* We will allow independent upgrade of the FRR component in order to resolve any
+  specific FRR bugs without upgrading MetalLB.
+* This enhancement assumes that each node’s MetalLB ‘speaker’ component can be
+  stopped (for an arbitrary time) and restarted without disruption of MetalLB as
+  a whole (To be confirmed)
+* This enhancement assumes that FRR will be deployed as a DaemonSet on the
+  Kubernetes cluster. It is also assumed that there is a stable/versioned
+  interface between this DaemonSet and MetalLB. Therefore, if any non-backwards
+  compatible changes are made in this interface, this will need to be managed in
+  the MetalLB ‘speaker’ component.
+* Upgrade/downgrade will remove a node from the MetalLB memberlist by unlabeling
+  the node. At this point the ‘speaker’ component on the node can be stopped and
+  the FRR component can be updated to the new version. It should be noted that
+  as FRR only peers with BGP agents outside the cluster, there is no requirement
+  that each node within the cluster maintains the same version or a compatible
+  version of FRR. However, it is required that each version of FRR is compatible
+  with the BGP agent with which it is peering. After FRR is updated, the
+  ‘speaker’ component can be restarted and the node can be labelled to add it to
+  the memberlist again.
+
+### Version Skew Strategy
+
+Version skew between versions of FRR should not be a concern to Kubernetes or
+MetalLB as long as FRR presents a stable interface to the MetalLB ‘speaker’
+component. This is because FRR will only peer with BGP agents outside of the
+cluster and not with each other.
+
+## Drawbacks
+
+This would couple MetalLB with the FRR community so that it would be somewhat
+dependent on this community for bug-fixes and further feature improvements.
+However, there would be the native BGP agent as a fallback for cluster
+administrators that did not want to deploy FRR.
```

Comment:
How would we handle cases where an upstream change to FRR is required by MetalLB but the FRR maintainers are either slow to implement it or they decide not to do that at all for whatever reason? And what about e.g. urgent security fixes?

- We can't simply fall back to the native stack because FRR has features the native stack doesn't.
- It doesn't make sense to ensure we only add FRR features that are supported natively because getting features "for free" was a central motivation for introducing FRR in the first place.
- If we end up adding functionality to MetalLB to keep up with relevant FRR features, perhaps it doesn't make sense to direct out efforts towards FRR in the first place.

I think that we can consider switching to FRR as the default BGP stack after some time has passed and we've gained confidence in the project, its release cadence etc. If we do that, maybe the problems above would be addressed.

Reply:
As FRR is an external dependency, I don't see a straightforward way to achieve this. A lot depends on the vibrancy of the project and the community as you have suggested.

From what I have seen/read, it does appear to be very active project under the umbrella of the Linux Foundation with regular releases and commits. The development workflow is documented [here](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#development-release-cycle) and does specify their process for back ports and release branches. Regarding security, they have a dedicated `security@lists.frrouting.org` address for handling security issues.

However, I suspect that any issues that we find would still be very relevant to their other use cases as what we are doing (compared to what other BGP deployments with FRR would be doing) is relatively straightforward. FWIW, their BGP PR/issue backlog seems quite active (https://github.com/FRRouting/frr/labels/bgp)

Reply:
I will also remove the native implementation as a fallback as I think we are now talking about its eventually retirement.